telemetry/core/src/main/scala/com/twitter/finagle/stats/buoyant/MetricsBucketedHistogram.scala

```diff
@@ -0,0 +1,311 @@
+    counts.synchronized {
+      if (num == 0) {
+        0L
+      } else if (counts(countsLength - 1) > 0) {
+        Int.MaxValue
+      } else {
+        var i = countsLength - 2 // already checked the last, start 1 before
+        while (i >= 0 && counts(i) == 0) {
+          i -= 1
+        }
+        if (i == 0) 0
+        else limitMidpoint(i)
+      }
+    }
+  }
+
+  /**
+   * The minimum value seen by calls to [[add]].
+   *
+   * @return 0 if no values have been added.
+   *         The returned value will be within
+   *         [[BucketedHistogram.DefaultErrorPercent]] of the actual value.
+   */
+  def minimum: Long = {
+    counts.synchronized {
+      if (num == 0) {
+        0L
+      } else {
+        var i = 0
+        while (i < countsLength && counts(i) == 0) {
+          i += 1
+        }
+        limitMidpoint(i)
+      }
+    }
+  }
+
+  /** Get the midpoint of bucket `i` */
+  private[this] def limitMidpoint(i: Int): Long = {
+    i match {
+      case 0 => 0
+      case _ if i >= limits.length => Int.MaxValue
+      case _ => (limits(i - 1).toLong + limits(i)) / 2
+    }
+  }
+
+  override def getQuantile(quantile: Double): Long =
+    percentile(quantile)
+
+  override def getQuantiles(quantiles: Array[Double]): Array[Long] = {
+    val ps = new Array[Long](quantiles.length)
+    var i = 0
+    while (i < ps.length) {
+      // Note: we could speed this up via just one pass over `counts` instead of
+      // of a pass per quantile.
+      // We could speed up calls to `percentile` by tracking the maximum
+      // bucket used during `add()`s to minimize how much of `counts` to scan.
+      ps(i) = percentile(quantiles(i))
+      i += 1
+    }
+    ps
+  }
+
+  /**
+   * The total of all the values seen by calls to [[add]].
+   */
+  def sum: Long = total
+
+  /**
+   * The number of values [[add added]].
+   */
+  def count: Long = num
+
+  /**
+   * The average, or arithmetic mean, of all values seen
+   * by calls to [[add]].
+   *
+   * @return 0.0 if no values have been [[add added]].
+   */
+  def average: Double = counts.synchronized {
+    if (num == 0) 0.0 else total / num.toDouble
+  }
+
+  /**
+   * Returns a seq containing nonzero values of the histogram.
+   * The sequence contains instances of BucketAndCount which are
+   * the bucket's upper and lower limits and a count of the number
+   * of times a value in range of the limits was added.
+   */
+  def bucketAndCounts: Seq[BucketAndCount] = counts.synchronized {
+    counts.zipWithIndex.collect {
+      case (count, idx) if count != 0 =>
+        // counts is 1 bucket longer than limits
+        // The last bucket of counts tracks added
+        // values greater than or equal to Int.MaxValue
+        val upperLimit = if (idx != limits.length) {
+          limits(idx)
+        } else Int.MaxValue
+        val lowerLimit = if (idx != 0) {
+          limits(idx - 1)
+        } else 0
+        BucketAndCount(lowerLimit, upperLimit, count)
+    }.toSeq
+  }
```

Comment:
Happy to make the change either way but it's not called that often, on the order of O(# of stats) per minute.